### PR TITLE
Design proposal regarding how we handle Server instantiation

### PR DESCRIPTION
**Pull request by mbaechler**
Follow up of discussion started here: https://github.com/linagora/james-project/pull/3531

It has been way harder than I initially thought.

  The design is the following:

  * Each Server define a Builder, there's no longer the need to figure out
    what are the requirement for a given Server as the builder is fluent enough

  * Each Server guice maven module defines an junit Extension to easily build a Server for
    testing. This extension relies on Server.Builder to retrieve configuration from caller
    and provide a phase to customize the Server definition using JamesServerBuilder

  * JamesServerBuilder still allows easy customization of the Server in the junit context
    for example by passing an Extension object

  As there's a lot of these Extensions invocationsm every refactoring costs a lot.
  By relying on Server.Builder, adding a configuration parameter should allow to
  only change these calls but no other abstraction.

### CODE REVIEW

**Comment by Arsnael** [comment]
Output build log lost

test this please

File: server/container/guice/cassandra-guice/src/test/java/org/apache/james/CassandraServerExtension.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+    public static class Builder {
+
+        private final JamesServerBuilder extensionBuilder;
+        private Optional<ExtensionCustomizer> customizer;
+
+        private Builder(JamesServerBuilder.ConfigurationProvider configuration) {
+            this.extensionBuilder = new JamesServerBuilder<Configuration>(
+                configuration,
+                c -> CassandraJamesServerMain.builder(c).build());
```
Comment (chibenwa) [inline server/container/guice/cassandra-guice/src/test/java/org/apache/james/CassandraServerExtension.java:53]: Can we find a better name for c ?

File: server/container/guice/cassandra-rabbitmq-guice/src/test/java/org/apache/james/CassandraRabbitMQServerExtension.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    public interface RequiresBlobConfiguration {
+        Builder blobStore(Function<BlobStoreConfiguration.Builder, BlobStoreConfiguration> configure);
+    }
+
+    @FunctionalInterface
+    public interface ExtensionCustomizer {
+        JamesServerBuilder customize(JamesServerBuilder builder);
+    }
+
+    public static class Builder {
```
Comment (chibenwa) [inline server/container/guice/cassandra-rabbitmq-guice/src/test/java/org/apache/james/CassandraRabbitMQServerExtension.java:52]: Do we need a builder given that we have by design a single `withSpecificParameters` (that do trigger a side effect)?

I would be in favor of handling withSpecificParameters via a Functional interface instead.

**Comment by Arsnael** [comment]
```
[6c5c74877b95977f0b9bbbfddeb82489101cb30e] [INFO] There is 1 error reported by Checkstyle 8.29 with checkstyle.xml ruleset.
[6c5c74877b95977f0b9bbbfddeb82489101cb30e] [ERROR] src/main/java/org/apache/james/JPAJamesServerMain.java:[22,8] (imports) UnusedImports: Unused import - java.util.Optional.
```

File: server/container/guice/cassandra-guice/src/test/java/org/apache/james/AuthenticatedCassandraJamesServerTest.java [hunk 3]
```diff
@@ -42,35 +44,43 @@
...
-            .extension(new DockerElasticSearchExtension())
-            .extension(cassandraExtension)
-            .disableAutoStart()
-            .server(configuration -> CassandraJamesServerMain.createServer(configuration)
-                .overrideWith(DOMAIN_LIST_CONFIGURATION_MODULE))
-            .overrideServerModule(binder -> binder.bind(ClusterConfiguration.class)
-                .toInstance(DockerCassandra.configurationBuilder(cassandraExtension.getCassandra().getHost())
-                    .username(CASSANDRA_USER)
-                    .password(VALID_PASSWORD)
-                    .useSsl()
```
Comment (rouazana) [inline server/container/guice/cassandra-guice/src/test/java/org/apache/james/AuthenticatedCassandraJamesServerTest.java:72]: you remove useSsl in a SslTest nested class?

File: server/container/guice/cassandra-rabbitmq-guice/src/main/java/org/apache/james/modules/blobstore/BlobStoreConfiguration.java [hunk 4]
```diff
@@ -37,24 +37,47 @@
...
 
-        public BlobStoreConfiguration enableCache() {
-            return new BlobStoreConfiguration(implementation, CACHE_ENABLED);
+        public CacheChoice objectStorage() {
+            return new CacheChoice(BlobStoreImplName.OBJECTSTORAGE);
         }
 
-        public BlobStoreConfiguration disableCache() {
-            return new BlobStoreConfiguration(implementation, !CACHE_ENABLED);
+        public BlobStoreConfiguration hybrid() {
```
Comment (rouazana) [inline server/container/guice/cassandra-rabbitmq-guice/src/main/java/org/apache/james/modules/blobstore/BlobStoreConfiguration.java:73]: is hybrid still a thing?

File: server/container/guice/cassandra-rabbitmq-guice/src/test/java/org/apache/james/CassandraRabbitMQJamesConfigurationTest.java [hunk 5]
```diff
@@ -1,115 +0,0 @@
...
-import static org.assertj.core.api.Assertions.assertThat;
-import static org.assertj.core.api.Assertions.assertThatThrownBy;
-
-import org.apache.james.server.core.MissingArgumentException;
-import org.apache.james.server.core.configuration.Configuration;
-import org.apache.james.server.core.configuration.Configuration.ConfigurationPath;
-import org.assertj.core.api.SoftAssertions;
-import org.junit.jupiter.api.Test;
-
-class CassandraRabbitMQJamesConfigurationTest {
```
Comment (rouazana) [inline server/container/guice/cassandra-rabbitmq-guice/src/test/java/org/apache/james/CassandraRabbitMQJamesConfigurationTest.java:31]: why do you remove this test class?

File: server/container/guice/cassandra-rabbitmq-ldap-guice/src/main/java/org/apache/james/CassandraRabbitMQLdapJamesServerMain.java [hunk 6]
```diff
@@ -24,33 +24,26 @@
...
+        Configuration configuration = Configuration.builder()
             .useWorkingDirectoryEnvProperty()
             .build();
+        BlobStoreConfiguration blobStoreConfiguration = BlobStoreConfiguration.parse(configuration);
 
         LOGGER.info("Loading configuration {}", configuration.toString());
-        GuiceJamesServer server = createServer(configuration)
-            .combineWith(new JMXServerModule());
+        GuiceJamesServer server = CassandraRabbitMQJamesServerMain.builder(configuration, blobStoreConfiguration)
+            .customize(s -> s
```
Comment (rouazana) [inline server/container/guice/cassandra-rabbitmq-ldap-guice/src/main/java/org/apache/james/CassandraRabbitMQLdapJamesServerMain.java:42]: server?

File: server/container/guice/memory-guice/src/test/java/org/apache/james/GuiceJamesServerStartUpCheckTest.java [hunk 7]
```diff
@@ -110,52 +114,51 @@ default void serverShouldStartSuccessfully(GuiceJamesServer server) throws Excep
...
-            .build();
+                .addBinding().to(TestBlobExportMechanismStartUpCheck.class)));
     }
 
     @Nested
     class WithNoStartUpCheck implements StartUpCheckSuccessContract {
 
         @RegisterExtension
-        JamesServerExtension jamesServerExtension = extensionBuilder().build();
+        JamesServerExtension jamesServerExtension = extensionBuilder(UnaryOperator.identity());
```
Comment (rouazana) [inline server/container/guice/memory-guice/src/test/java/org/apache/james/GuiceJamesServerStartUpCheckTest.java:147]: could there be an `extensionBuilder()` defaulting to `identitiy()`?

**Comment by chibenwa** [comment]
```
[6c5c74877b95977f0b9bbbfddeb82489101cb30e] [ERROR] src/main/java/org/apache/james/JPAJamesServerMain.java:[22,8] (imports) UnusedImports: Unused import - java.util.Optional.
```

**Comment by chibenwa** [comment]
This had been more or less implemented.

Thanks for the suggestion!

**Pull request closed without merging**